File: packages/shared/src/money.ts
/**
 * Money helpers. We store money as integer minor units ("paise" for INR,
 * "cents" generally) everywhere to avoid floating-point errors, and only format
 * to a decimal string at the edges. Defaults are India-first (INR, en-IN), which
 * gives the ₹ symbol and lakh/crore digit grouping.
 */

export function formatMoney(minorUnits: number, currency = 'INR', locale = 'en-IN'): string {
  return new Intl.NumberFormat(locale, { style: 'currency', currency }).format(minorUnits / 100);
}

/** Parse a user-entered decimal amount (e.g. "1299.00") into integer minor units. */
export function toCents(amount: number | string): number {
  const value = typeof amount === 'string' ? Number(amount) : amount;
  if (!Number.isFinite(value)) throw new Error(`Invalid money amount: ${amount}`);
  return Math.round(value * 100);
}

export function fromCents(minorUnits: number): number {
  return minorUnits / 100;
}

File: mobile/lib/core/money.dart
import 'package:intl/intl.dart';

/// Money is integer minor units (paise) end-to-end; format only at the edge,
/// India-first (₹, en-IN lakh/crore grouping).
final NumberFormat _inr = NumberFormat.currency(
  locale: 'en_IN',
  symbol: '₹',
  decimalDigits: 2,
);

String formatMoney(int minorUnits, {String currency = 'INR'}) {
  final double major = minorUnits / 100;
  if (currency == 'INR') return _inr.format(major);
  return NumberFormat.currency(locale: 'en_IN', name: currency).format(major);
}

/// Human "time left" for a desired-by date, mirroring the web client.
String formatHorizon(DateTime desiredBy) {
  final int days = desiredBy.difference(DateTime.now()).inHours ~/ 24;
  if (days < 0) return 'past due';
  if (days == 0) return 'due today';
  if (days < 31) return '$days day${days == 1 ? '' : 's'} left';
  final int months = (days / 30).round();
  return '~$months month${months == 1 ? '' : 's'} left';
}
